Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -0,0 +1,149 @@
+/-
+Copyright (c) 2024 Judith Ludwig, Florent Schaffhauser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Judith Ludwig, Florent Schaffhauser, Yunzhou Xie, Jujian Zhang
+-/
+
+import Mathlib.RingTheory.Flat.Basic
+
+/-!
+# Faithfully flat modules
+
+A module `M` over a commutative ring `R` is *faithfully flat* if it is flat and `IM ≠ M` whenever
+`I` is a maximal ideal of `R`.
+
+## Main declaration
+
+- `Module.FaithfullyFlat`: the predicate asserting that an `R`-module `M` is faithfully flat.
+
+## Main theorems
+
+- `Module.FaithfullyFlat.iff_flat_and_proper_ideal`: an `R`-module `M` is faithfully flat iff it is
+  flat and for all proper ideals `I` of `R`, `I • M ≠ M`.
+- `Module.FaithfullyFlat.iff_flat_and_rTensor_faithful`: an `R`-module `M` is faithfully flat iff it
+  is flat and tensoring with `M` is faithful, i.e. `N ≠ 0` implies `N ⊗ M ≠ 0`.
+- `Module.FaithfullyFlat.iff_flat_and_lTensor_faithful`: an `R`-module `M` is faithfully flat iff it
+  is flat and tensoring with `M` is faithful, i.e. `N ≠ 0` implies `M ⊗ N ≠ 0`.
+
+- `Module.FaithfullyFlat.self`: the `R`-module `R` is faithfully flat.
+
+-/
+
+universe u v
+
+open TensorProduct
+
+namespace Module
+
+variable (R : Type u) (M : Type v) [CommRing R] [AddCommGroup M] [Module R M]
+
+/--
+A module `M` over a commutative ring `R` is *faithfully flat* if it is flat and,
+for all `R`-module homomorphism `f : N → N'` such that `id ⊗ f = 0`, we have `f = 0`.
+-/
+@[mk_iff] class FaithfullyFlat : Prop where
+  flat : Module.Flat R M := by infer_instance
+  submodule_ne_top :  ∀ ⦃m : Ideal R⦄ (_ : Ideal.IsMaximal m), m • (⊤ : Submodule R M) ≠ ⊤
+
+namespace FaithfullyFlat
+
+attribute [instance] FaithfullyFlat.flat
+
+instance self (R : Type u) [CommRing R] : FaithfullyFlat R R where
+  submodule_ne_top m h r := Ideal.eq_top_iff_one _ |>.not.1 h.ne_top <| by
+    simpa using show 1 ∈ (m • ⊤ : Ideal R) from r.symm ▸ ⟨⟩
+
+section proper_ideal
+
+lemma iff_flat_and_proper_ideal :
+    FaithfullyFlat R M ↔
+    (Flat R M ∧ ∀ (I : Ideal R), I ≠ ⊤ → I • (⊤ : Submodule R M) ≠ ⊤) := by
+  rw [faithfullyFlat_iff]
+  refine ⟨fun ⟨flat, h⟩ => ⟨flat, fun I hI r => ?_⟩, fun h => ⟨h.1, fun m hm => h.2 _ hm.ne_top⟩⟩
+  obtain ⟨m, hm, le⟩ := I.exists_le_maximal hI
+  exact h hm <| eq_top_iff.2 <| show ⊤ ≤ m • ⊤ from r ▸ Submodule.smul_mono le (by simp [r])
+
+end proper_ideal
+
+section faithful
+
+attribute [-simp] Ideal.Quotient.mk_eq_mk in
+lemma iff_flat_and_rTensor_faithful :
+    FaithfullyFlat R M ↔
+    (Flat R M ∧
+      ∀ (N : Type max u v) [AddCommGroup N] [Module R N],
```

Comment:
Maybe you can use `∀ ⦃N : Type ...'⦄`, but as you prefer.

Does this hold with `N` in an arbitrary universe? (Of course we only want the `→` direction with such an `N`.)

Reply:
I have added instance `Module.FaithfullyFlat.lTensor_nontrivial` that works for `N` in arbitrary `universes`

Reply:
For the `⦃N : Type ...'⦄`, I think an explicit `N` works better. When dealing with exactness, i.e. `⦃N1 N2 N3 : Type ...'⦄`, I think I will use `⦃⦄`